### PR TITLE
Enhance serial interfaces

### LINKED ISSUE #181: Implement timeout for send and receive method for serial interface

### PR DESCRIPTION
**Pull request by boris-wenzlaff**
- Rename serial can to simple serial
- Adjust simple serial and slcan interface
- Add generic interface test class
- Add documentation for generic interface test class
- Add test cases for slcan
- Enhance documentation for slcan
- Implementation of timeouts for simple serial issue (Closes #181)

### CODE REVIEW

**Comment by felixdivo** [comment]
Wow, really nice! The `GenericInterfaceTest` class could be used in other interferes as well.

**Comment by yegorich** [comment]
@boris-wenzlaff great rework! I have a minor improvement suggestion. Could you use serial.serial_for_url() instead of serial.Serial()? This would enable support for network based CAN converters.

**Comment by yegorich** [comment]
@boris-wenzlaff I have tested your PR and encountered a decoding issue in Debian 9, Python 3.5.3:

```
root@onrisc:/usr/local/src/python-can# python3 /usr/local/bin/canslcan.py
Traceback (most recent call last):
  File "/usr/local/bin/canslcan.py", line 16, in <module>
    msg = bus.recv(None)
  File "/usr/local/lib/python3.5/dist-packages/python_can-2.1.1.dev0-py3.5.egg/can/interfaces/serial/slcan.py", line 185, in recv
    read_line = self.serial_port.readline()
  File "/usr/lib/python3.5/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 0: ordinal not in range(128)

```

I'm using native CAN interface to send data via: `cangen can0 -e -L i -v -v -v`

**Comment by yegorich** [comment]
By the way, what about adding an option to enable handshake? I don't know whether XON/OFF makes much sense but RTS/CTS is really a must for USB-to-CAN converter as some of them run with 3Mbit/s on the serial side.

**Comment by felixdivo** [comment]
Wouldn't the `Howto` page fit nicely in `doc/Development.rst`?

And why exactly did you rename the interface to `simpleserial` after you changed it. It breaks compatibility with older versions of this library.

**Comment by yegorich** [comment]
Any progress on this PR?

@boris-wenzlaff what about send v2 with serial_for_url and hw handshake?

File: can/interface.py [hunk 1]
```diff
@@ -28,19 +28,19 @@
...
-    'iscan':            ('can.interfaces.iscan',            'IscanBus'),
-    'virtual':          ('can.interfaces.virtual',          'VirtualBus'),
-    'neovi':            ('can.interfaces.ics_neovi',        'NeoViBus'),
-    'vector':           ('can.interfaces.vector',           'VectorBus'),
-    'slcan':            ('can.interfaces.slcan',            'slcanBus')
+    'kvaser':           ('can.interfaces.kvaser',             'KvaserBus'),
+    'socketcan_ctypes': ('can.interfaces.socketcan',          'SocketcanCtypes_Bus'),
+    'socketcan_native': ('can.interfaces.socketcan',          'SocketcanNative_Bus'),
+    'simpleserial':     ('can.interfaces.serial.simpleserial','SimpleSerialBus'),
+    'slcan':            ('can.interfaces.serial.slcan',       'SlcanBus')
```
Comment (felixdivo) [inline can/interface.py]: There is a comma missing at the end of this line. I must've missed it while merging some conflict.

**Comment by boris-wenzlaff** [comment]
@yegorich In the moment I'am short of time. I will later finish the branch.

**Comment by felixdivo** [comment]
@boris-wenzlaff Any updates on this?

**Comment by felixdivo** [comment]
@boris-wenzlaff Do you need any help with this PR?

**Comment by felixdivo** [comment]
I guess @boris-wenzlaff will not finish this ... It has also gone a lot out of sync with the current branch.

Is there someone willing to pick this up, and create a PR with the relevant changes?

**Comment by felixdivo** [comment]
> I guess @boris-wenzlaff will not finish this

What do we do with this wasteland? I liked the generic test class for busses. Se wo could port (~=copy) this:

- `doc/howto.rst` (!)
- `test/interface_test.py`(!)
- and maybe `test/interface_slcan_test.py` or `test/interface_simpleserial_test.py` can be used as a replacement for the current `test/serial_test.py`
- a bit of documentation from `doc/interfaces/slcan.rst`

But I don't have the time to do that. The remaining stuff seems quite out of sync ...

**Comment by karlding** [comment]
Closing. Please reopen and rebase if needed.

**Pull request closed without merging**